### ISSUE TITLE
build test failure on Debian

### ISSUE THREAD

**Issue by rickysarraf** [libstorage/libstoragemgmt#153]
I am working on refreshing libstoragemgmt and ran into the following build (test) failure.

In the past, I've been building it with the following option, which now fails on 1.3.1. 

```
 override_dh_auto_configure:
        autoreconf -i -f
       ./configure --prefix=/usr --disable-build-c-unit
```

```
executing: chmod 0777 /tmp/2072/logs/
executing: libtool --warning=none --mode install install /build/libstoragemgmt-1.3.1/daemon/lsmd /tmp/2072/bin/
libtool: install: install /build/libstoragemgmt-1.3.1/daemon/lsmd /tmp/2072/bin/lsmd
executing: libtool --warning=none --mode install install /build/libstoragemgmt-1.3.1/c_binding/libstoragemgmt.la /tmp/2072/c_libs/
libtool: install: install /build/libstoragemgmt-1.3.1/c_binding/.libs/libstoragemgmt.so.1.3.1 /tmp/2072/c_libs/libstoragemgmt.so.1.3.1
libtool: install: (cd /tmp/2072/c_libs && { ln -s -f libstoragemgmt.so.1.3.1 libstoragemgmt.so.1 || { rm -f libstoragemgmt.so.1 && ln -s libstoragemgmt.so.1.3.1 libstoragemgmt.so.1; }; })
libtool: install: (cd /tmp/2072/c_libs && { ln -s -f libstoragemgmt.so.1.3.1 libstoragemgmt.so || { rm -f libstoragemgmt.so && ln -s libstoragemgmt.so.1.3.1 libstoragemgmt.so; }; })
libtool: install: install /build/libstoragemgmt-1.3.1/c_binding/.libs/libstoragemgmt.lai /tmp/2072/c_libs/libstoragemgmt.la
libtool: install: install /build/libstoragemgmt-1.3.1/c_binding/.libs/libstoragemgmt.a /tmp/2072/c_libs/libstoragemgmt.a
libtool: install: chmod 644 /tmp/2072/c_libs/libstoragemgmt.a
libtool: install: ranlib /tmp/2072/c_libs/libstoragemgmt.a
executing: cp /build/libstoragemgmt-1.3.1/python_binding/lsm/.libs/_clib.so /tmp/2072/python_modules//lsm/_clib.so
executing: chrpath -d /tmp/2072/python_modules//lsm/_clib.so
executing: find /build/libstoragemgmt-1.3.1/python_binding/lsm/ -maxdepth 1 -type f -name *.py -exec install -D {} /tmp/2072/python_modules//lsm/ \;
executing: find /build/libstoragemgmt-1.3.1/python_binding/lsm/external/ -maxdepth 1 -type f -name *.py -exec install -D {} /tmp/2072/python_modules//lsm/external \;
executing: find /build/libstoragemgmt-1.3.1/tools/lsmcli/ -maxdepth 1 -type f -name *.py -exec install -D {} /tmp/2072/python_modules//lsm/lsmcli/ \;
executing: install -D /build/libstoragemgmt-1.3.1/tools/lsmcli/lsmcli /tmp/2072/bin//lsmcli
executing: libtool --warning=none --mode install install /build/libstoragemgmt-1.3.1/test/tester /tmp/2072/bin//tester
libtool: install: install /build/libstoragemgmt-1.3.1/test/tester /tmp/2072/bin//tester
install: cannot stat '/build/libstoragemgmt-1.3.1/test/tester': No such file or directory
Fail exit[1]: libtool --warning=none --mode install install /build/libstoragemgmt-1.3.1/test/tester /tmp/2072/bin//tester
============ Dumping log BEGIN =====================
============ Dumping log END   =====================
Base folder is '/tmp/2072', please investigate
Base folder /tmp/2072 is kept for investigation
FAIL test/runtests.sh (exit status: 1)

============================================================================
Testsuite summary for libstoragemgmt 1.3.1
============================================================================
# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
```

If I enable the tests, the build progresses futher, and then fails with the following.

[libstoragemgmt_1.3.1-1_amd64.build.txt](https://github.com/libstorage/libstoragemgmt/files/268261/libstoragemgmt_1.3.1-1_amd64.build.txt)

If it is not important, I'd like to disable tests at build time. What is your recommendation as the upstream for this project ?

**Comment by joehandzik** [comment]
From a quick look at the logs, this would appear to be our problem:

tester.c:3367:F:Basic:test_local_disk_vpd83_search:0: lsm_local_disk_vpd83_search(): Expecting LSM_ERR_OKwhen no argument is NULL
tester.c:3295:F:Basic:test_local_disk_list:0: lsm_local_disk_list() failed as 1
tester.c:3834:F:Basic:test_local_disk_ident_led:0: lsm_local_disk_list() failed as 1
tester.c:3841:F:Basic:test_local_disk_fault_led:0: lsm_local_disk_list() failed as 1
Fail exit[1]: /tmp/22269/bin//tester

I think we want the tests if at all possible, though we'll need to take a look at why this is failing to see if it matters. Thanks for bringing it to our attention!

**Comment by tasleson** [comment]
This failure could be related to the one Gris just posted a fix for.  I will try it out on a debian VM and see if I can reproduce and if Gris's change corrects it.  We need to roll a 1.3.2 release so we can address this issue too.  As for bypassing tests during builds, we should have a configure argument that does just that.  Running the local tests isn't a requirement, but it is certainly helpful for ensuring the build is in a somewhat workable state.

@rickysarraf What version of debian are you building on?

**Comment by joehandzik** [comment]
@tasleson Good point, the build is 1.3.1.

**Comment by rickysarraf** [comment]
This is on Debian unstable/sid

sent fr0m a $martph0nè, excuse typ0s

**Comment by rickysarraf** [comment]
@tasleson  Yes. Fixing the current option (to disable tests) will suffice for this bug report.

For Debian, I've had the tests disabled for previous releases too because IIRC the tests had failed.

**Comment by tasleson** [comment]
@rickysarraf OK, some observations from looking at the log file you supplied:

This is what is passed to configure
`717 ./configure --prefix=/usr`

Where your build is calling `make check`
```
1051    dh_auto_test
1052	make -j1 check VERBOSE=1
```

I looked at `--disable-build-c-unit`, basically it bypasses the traversal of the `test` sub directory, so the C unit test doesn't get built.  If you actually try to run `make check` it will fail with:

`install: cannot stat '/home/tasleson/libstoragemgmt-1.3.1/test/tester': No such file or directory`

I did a `make check` on debian 8 (Don't have sid at the moment) and it passes clean when running as root and non-root user.  From the errors in the log file, it looks like we are probably running into something with the fake root environment, but not 100% sure.  I'm not familiar with the build process on debian so at the moment I'm unable to replicate.

So the questions I have are:
1. If you don't want to run the tests why is your build doing a `make check`? 
2. Would it be possible for you to try out the latest from git and see if it corrects the test case failure you are running into?

I guess I'm hesitant to add a configure argument who's sole purpose would be to allow someone to run `make check` which would basically be a no-op instead of actually running tests.  Gris added an option to disable valgrind tests which is helpful, but that's when you want to run a subset of the tests, not run none of them.  To get the packages to build on Fedora & EPEL I did remove the `make check` as we were running into a valgrind failure in `ld.so` on arm arch, which has nothing to do with libStorageMgmt.

**Comment by cathay4t** [comment]
We might add configure option `--without test` to suppress all build requirement for test cases like `chrpath, ps, check-devel, perl, wc, valgrind`.

**Comment by rickysarraf** [comment]
@tasleson Following is what my build config is. Anything else is picked from the defaults. The only exception is that in the following example, I dropped the '-disable-build-c-unit' option because it is failing. All previous [builds](https://buildd.debian.org/status/package.php?p=libstoragemgmt) in Debian have it disabled.

```
rrs@learner:~/.rrs-home/Community/Packaging/libstoragemgmt (master)$ cat debian/rules 
#!/usr/bin/make -f

# Uncomment this to turn on verbose mode.
#export DH_VERBOSE=1

# For Hardening
export DEB_BUILD_MAINT_OPTIONS = hardening=+all
DPKG_EXPORT_BUILDFLAGS = 1
include /usr/share/dpkg/buildflags.mk

%:
	dh $@ --with python2,autoreconf

override_dh_auto_configure:
	autoreconf -i -f
	./configure --prefix=/usr

2016-05-18 / 14:20:56 ♒♒♒  ☺  
```

So, that `make check` must be part of (your) default build config. As mentioned in the debian/rules snippet above, I'm not triggering it.

@cathay4t Yes. That'll help. Because we build on much more architectures than any other distribution. And not all of them have the hardware resources.

**Comment by rickysarraf** [comment]
I just dug my emails and realized that this bug was existent back in 2013 too. Just that, back then, the '--disable-build-c-unit' worked fine in disabling the tests.

```
Try doing a configure with --disable-build-c-unit and let me know if
that helps.

Regards,
Tony

On 06/16/2013 04:09 AM, Ritesh Raj Sarraf wrote:
> Tony,
> 
> I am trying to package verison 0.0.20.
> 
> 
> I get the following build failure:
> 
> Making check in test
> make[2]: Entering directory `/tmp/buildd/libstoragemgmt-0.0.20/test'
> make  tester
> make[3]: Entering directory `/tmp/buildd/libstoragemgmt-0.0.20/test'
>   CC       tester-tester.o
>   CCLD     tester
> /usr/bin/ld:
> /usr/lib/gcc/x86_64-linux-gnu/4.7/../../../x86_64-linux-gnu/libcheck_pic.a(check_pack.o):
> undefined reference to symbol '__pthread_unregister_cancel@@GLIBC_2.3.3'
> /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing
> from command line
> collect2: error: ld returned 1 exit status
> make[3]: *** [tester] Error 1
> make[3]: Leaving directory `/tmp/buildd/libstoragemgmt-0.0.20/test'
> make[2]: *** [check-am] Error 2
> make[2]: Leaving directory `/tmp/buildd/libstoragemgmt-0.0.20/test'
> make[1]: *** [check-recursive] Error 1
> make[1]: Leaving directory `/tmp/buildd/libstoragemgmt-0.0.20'
> dh_auto_test: make -j1 check returned exit code 2
> make: *** [build] Error 29
> dpkg-buildpackage: error: debian/rules build gave error exit status 2
> E: Failed autobuilding of package
> I: unmounting /var/cache/apt/archives/ filesystem
> I: unmounting dev/pts filesystem
> I: unmounting run/shm filesystem
> I: unmounting proc filesystem
> I: cleaning the build env
> I: removing directory /tmp/876 and its subdirectories
> gbp:error: Couldn't run '/home/rrs/bin/gbp-pbuilder':
> /home/rrs/bin/gbp-pbuilder returned 1
> 
> Any help?
> 
```

**Comment by cathay4t** [comment]
Thanks for mention that.
I will replace buggy `--disable-build-c-unit` option with `without-test` in coming PR.

**Comment by cathay4t** [comment]
New configure option added in PR #154, please review.

**Comment by tasleson** [comment]
@rickysarraf Please see: https://manpages.debian.org/cgi-bin/man.cgi?query=dh_auto_test

**Comment by tasleson** [comment]
@rickysarraf @cathay4t I've merged PR #154.  However, in this case I would have preferred to understand the test failure before blindly disabling tests on debian.  Furthermore, the debian build environment appears to be running `make check` automatically and can be controlled with

```text
If the DEB_BUILD_OPTIONS environment variable contains nocheck, no
tests will be performed.
```

I will be creating a 1.3.2 release today.

**Comment by rickysarraf** [comment]
@tasleson Thanks. Debhelper 9 and above simplifies packaging in Debian. On any standard build environment (including your's using autotools), packaging can be trimmed down to a mere 2 liner. In such cases, it performs the defaults.

Like [this](http://anonscm.debian.org/cgit/apt-offline/apt-offline.git/tree/debian/rules?h=debian) for example.

```
rrs@learner:~/.rrs-home/devel/apt-offline/apt-offline (debian)$ cat debian/rules 
#!/usr/bin/make -f
%:
	dh $@ --with python2

2016-05-18 / 22:37:42 ♒♒♒  ☺  
```
